File: Data Structures & Algorithms/validate-parentheses/submission-13.py
class Solution:
    def isValid(self, s: str) -> bool:
        close_to_open = {')': '(', '}': '{', ']': '['}
        stack = []

        if len(s) % 2:
            return False

        for p in s:
            if p not in close_to_open:
                stack.append(p)
            elif stack and close_to_open[p] == stack[-1]:
                stack.pop()
            else:
                return False
        
        return not stack
        

File: Data Structures & Algorithms/daily-temperatures/submission-13.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<pair<int, int>> stack;
        vector<int> results(temperatures.size(), 0);

        for (int i = 0; i < temperatures.size(); i++) {
            int current = temperatures[i];

            while (!stack.empty() && current > stack.top().second) {
                auto pair = stack.top();
                results[pair.first] = i - pair.first;
                stack.pop();
            }

            stack.push({i, current});
        }
        
        return results;
    }
};
